src/4_dbscan.md

```diff
@@ -0,0 +1,51 @@
+
+Comparatively, KMeans will take into account each point in the dataset, which means outliers can negatively affect the local optimal location for a given cluster's centroid in order to accommodate them. 
+
+### Using DSBCAN with `linfa`
+
+Compared to 
+```rust,no_run
+{{#include ../code/examples/dbscan.rs:libraries}}
+```
+
+Instead of having a several higher-density clusters different areas, we'll take advantage of DBSCAN's ability to follow spatially-contiguous non-localized clusters by building our data out of both filled and hollow circles, with some random noise tossed in as well. The end goal will be to re-find each of these clusters, and exclude some of the noise!
+
+```rust,no_run
+{{#include ../code/examples/dbscan.rs:create_circles}}
+```
+Compared to `linfa`'s KMeans algorithm, the DBSCAN implementation is able to operate directly on a ndarray `Array2` data structure, so there's no need to convert it into the `linfa`-native `Dataset` type first. It's also worth pointing out that choosing the chosen parameters often take some experimentation and tuning before they produce results that actually make sense. This is one of the areas where data visualization can be really valuable; it is helpful in developing some physical intuition about your data set and understand how your choice of hyperparameters will affect the results produced by the algorithm.  
+
+```rust,no_run
+{{#include ../code/examples/dbscan.rs:create_and_run_model}}
+```
+We'll skip over setting up `ChartBuilder` struct and drawing areas from the `plotters` crate, since it's exactly the same as in the [KMeans](./3_kmeans.md) example. 
+
+Remember how we mentioned DBSCAN is an algorithm that can exclude noise? That's particularly important for the pattern-matching in this case, since we're almost guaranteed to end up with some values that don't fit nicely into any of our expected clusters. Since we generated an artificial dataset, we know the number of clusters that should be generated, and where they're located. However, that won't always be the case. In that situation, we could instead examine the number of clusters afterwards, create a colormap using custom RGB colors which matches the highest number of clusters, and plot it that way.
+
+```rust,no_run
+{{#include ../code/examples/dbscan.rs:plot_points}}
+```
+
+As a result, we then get the following chart, where each cluster is uniquely identified, and some of the random noise associated with the dataset is discarded.  
+
+![DBSCAN](dbscan.png)
```

Comment:
it would also be good to have some reference to the original paper

Reply:
I think there's I linked to the original paper in the heading underneath "What is DBSCAN?", somewhere around line 5